docs/contributing/testing-guidelines/unit-testing.md

```diff
@@ -0,0 +1,165 @@
+
+See below for an example of using `ament_add_ros_isolated_gtest` with `colcon test`.
+All other tests follow a similar pattern.
+
+## Create a unit test with gtest
+
+In `my_cool_pkg/test`, create the `gtest` code file `test_my_cool_pkg.cpp`:
+
+```cpp
+#include "gtest/gtest.h"
+#include "my_cool_pkg/my_cool_pkg.hpp"
+TEST(TestMyCoolPkg, TestHello) {
+  EXPECT_EQ(my_cool_pkg::print_hello(), 0);
+}
+```
+
+For more examples of `gtest` features, see the [gtest repo](https://github.com/google/googletest).
+
+In `package.xml`, add the following line:
+
+```xml
+<test_depend>ament_cmake_ros</test_depend>
+```
+
+Next add an entry under `BUILD_TESTING` in the `CMakeLists.txt` to compile the test source files:
+
+```cmake
+if(BUILD_TESTING)
+
+  ament_add_ros_isolated_gtest(test_my_cool_pkg test/test_my_cool_pkg.cpp)
+  target_link_libraries(test_my_cool_pkg ${PROJECT_NAME})
+...
+endif()
+```
+
+This automatically links the test with the default main function provided by `gtest`.
+The code under test is usually in a different CMake target (`${PROJECT_NAME}` in the example) and its shared object for linking need to be added.
+
+To register a new `gtest` item, wrap the test code with the macro `TEST ()`.
+`TEST ()` is a predefined macro that helps generate the final test code,
+and also registers a `gtest` item to be available for execution.
+The test case name should be in CamelCase, since gtest inserts an underscore between the fixture name and the class case name when creating the test executable.
+
+`gtest/gtest.h` also contains predefined macros of `gtest` like `ASSERT_TRUE(condition)`,
+`ASSERT_FALSE(condition)`, `ASSERT_EQ(val1,val2)`, `ASSERT_STREQ(str1,str2)`, `EXPECT_EQ()`, etc.
+`ASSERT_*` will abort the test if the condition is not satisfied,
+while `EXPECT_*` will mark the test as failed but continue to next test condition.
+More information about `gtest` can be found in the [gtest repo](https://github.com/google/googletest).
+
+In the demo `CMakeLists.txt`, `ament_add_ros_isolated_gtest` is a predefined macro in `ament_cmake_ros` that helps simplify adding `gtest` code.
+Details can be viewed in [ament_add_gtest.cmake](https://github.com/ros2/ament_cmake_ros/tree/master/ament_cmake_ros/cmake).
+
+## Build test
+
+<!-- cspell:ignore Testfile -->
+
+By default, all necessary test files (`ELF`, `CTestTestfile.cmake`, etc.) are compiled by `colcon`:
+
+```bash
+cd ~/workspace/
+colcon build --packages-select my_cool_pkg
+```
+
+Test files are generated under `~/workspace/build/my_cool_pkg`.
+
+## Run test
+
+To run all tests for a specific package, call:
+
+```bash
+colcon test --packages-select my_cool_pkg
+
+Starting >>> my_cool_pkg
+Finished <<< my_cool_pkg [7.80s]
+
+Summary: 1 package finished [9.27s]
+```
+
+The test command output contains a brief report of all the test results.
+
+To get job-wise information of all executed tests, call:
+
+```bash
+colcon test-result --all
```

Comment:
If there are some output logs, I remember it's recommended to add the `$` mark to distinguish the command and output.
```suggestion
$ colcon test-result --all
```

Reply:
Also, you can consider replacing `bash` with `sh-session` to highlight correctly.

Reply:
I'm sorry I couldn't find a good reference, but this is related code.
https://github.com/github/linguist/blob/2e1c938e8ca474e03161c2fa010650b2331223ca/lib/linguist/languages.yml#L6032

Reply:
Thank you. Yeah, Markdown has some dialects. :cry: